### PR TITLE
Moar `sg migration` niceness

### PR DESCRIPTION
**Pull request by tjdevries**
By the end of this PR, I hope to:

1. Not require `migrate` to be installed locally (the squash things scare me and I won't touch them. I just moved to new file)
2. `up` and `down` use connection to DB
3. `fixup` now resets database to correct location, fixes up conflicting migrations, and then applies new migrations, so your DB is in the right spot.

1 & 2 are mostly done.

Status / Info for review:
- `migration.go` got split into two areas
  - `squash/squash.go` which contains all the code for squashing migrations. I don't really understand this code, but I pulled it all out of `migration.go` because it doesn't connect to the db the same way, has a bunch of names that similar and confused me every time and because I didn't 
  - `migration/migration.go` is now just about doing actual migrations in the database. It now connects in Go via the same library we use in prod, as opposed to using the shell scripts, which should let us do a lot cooler things.

Otherwise I'm just trying to leave other things as untouched as I can for future refactors.

Sorry if any of  this is overkill, I was also learning lots of things about how to structure Go projects as I was doing it.

### CODE REVIEW

**Comment by tjdevries** [comment]
Status:

```
[I] ➜ go build && ./sg migration fixup -db codeintel -run=false

Checking for conflicting migrations in 'codeintel'...
 Latest migration in the main branch  : 1000000017 (lsif_documentation_path_info)
 Latest migration in your local branch: 1000000019 (does_literally_nothing_12)
 Current migration version in database: 1000000018

 Pending Operations:
   OpTempfile        : 1000000017 lsif_documentation_path_info
   OpDatabaseDown    : Migrate down past 1000000018 (does_literally_nothing)
   OpDatabaseDown    : Migrate down past 1000000017 (nothing)
   OpReverseTempfile : 1000000017 lsif_documentation_path_info
   OpFileFixup       : Conflicting migration main:lsif_documentation_path_info local:nothing
   OpFileShift       : Shift migration 'does_literally_nothing' from 1000000018 -> 1000000019
   OpFileShift       : Shift migration 'does_literally_nothing_12' from 1000000019 -> 1000000020
   OpDatabaseSync    : Sync to latest
```

When actually running:

```
[I] ➜ go build && ./sg migration fixup -db codeintel -run=true

Checking for conflicting migrations in 'codeintel'...
 Latest migration in the main branch  : 1000000017 (lsif_documentation_path_info)
 Latest migration in your local branch: 1000000019 (does_literally_nothing_12)
 Current migration version in database: 1000000018

 Pending Operations:
   OpTempfile        : 1000000017 lsif_documentation_path_info
   OpDatabaseDown    : Migrate down past 1000000018 (does_literally_nothing)
   OpDatabaseDown    : Migrate down past 1000000017 (nothing)
   OpReverseTempfile : 1000000017 lsif_documentation_path_info
   OpFileFixup       : Conflicting migration main:lsif_documentation_path_info local:nothing
   OpFileShift       : Shift migration 'does_literally_nothing' from 1000000018 -> 1000000019
   OpFileShift       : Shift migration 'does_literally_nothing_12' from 1000000019 -> 1000000020
   OpDatabaseSync    : Sync to latest
 ✅ 1000000017_lsif_documentation_path_info.up.sql -> _1000000017_lsif_documentation_path_info.up.sql
 ✅ 1000000017_lsif_documentation_path_info.down.sql -> _1000000017_lsif_documentation_path_info.down.sql
 1000000018/d does_literally_nothing (57.51514ms)
 1000000017/d nothing (36.831654ms)
 ✅ _1000000017_lsif_documentation_path_info.up.sql -> 1000000017_lsif_documentation_path_info.up.sql
 ✅ _1000000017_lsif_documentation_path_info.down.sql -> 1000000017_lsif_documentation_path_info.down.sql
 Changing migration: 1000000017 lsif_documentation_path_info
 ✅ 1000000017_nothing.up.sql -> 1000000018_nothing.up.sql
 ✅ 1000000017_nothing.down.sql -> 1000000018_nothing.down.sql
 ✅ 1000000018_does_literally_nothing.up.sql -> 1000000019_does_literally_nothing.up.sql
 ✅ 1000000018_does_literally_nothing.down.sql -> 1000000019_does_literally_nothing.down.sql
 ✅ 1000000019_does_literally_nothing_12.up.sql -> 1000000020_does_literally_nothing_12.up.sql
 ✅ 1000000019_does_literally_nothing_12.down.sql -> 1000000020_does_literally_nothing_12.down.sql
 Applying migrations to local database
 1000000017/u lsif_documentation_path_info (87.128647ms)
 1000000018/u nothing (118.847003ms)
 1000000019/u does_literally_nothing (142.861901ms)
 1000000020/u does_literally_nothing_12 (177.441402ms)
```

And then, since DB is cleaned:

```
Checking for conflicting migrations in 'codeintel'...
 Latest migration in the main branch  : 1000000017 (lsif_documentation_path_info)
 Latest migration in your local branch: 1000000020 (does_literally_nothing_12)
 Current migration version in database: 1000000020
 ✅ Migrations and database are synced.
```

File: dev/sg/internal/db/db.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import (
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/sourcegraph/sourcegraph/dev/sg/root"
+)
+
+func getFsForPath(path string) fs.FS {
```
Comment (efritz) [inline dev/sg/internal/db/db.go:12]: Looks dumb, but `getFSForPath` is the idiomatic name.

File: dev/sg/internal/migration/migration.go [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+	// Negative ints mean that we're going down.
+	if n != nil {
+		*n = -1 * *n
+	}
+
+	return doRunMigrations(database, n, out, "down", func(m *migrate.Migrate) error { return m.Down() })
+}
+
+// RunFixup will run the fixup command.
+// run determines whether to actually run the migrations in your local database or not
```
Comment (efritz) [inline dev/sg/internal/migration/migration.go:45]: ?

File: dev/sg/output.go [hunk 3]
```diff
@@ -3,10 +3,16 @@ package main
...
+	"github.com/sourcegraph/sourcegraph/dev/sg/internal/squash"
 	"github.com/sourcegraph/sourcegraph/lib/output"
 )
 
 var out = output.NewOutput(os.Stdout, output.OutputOpts{
 	ForceColor: true,
 	ForceTTY:   true,
 })
+
+// TODO: Sad Hacks
```
Comment (efritz) [inline dev/sg/output.go:15]: Maybe export `out` instead and reverse the dependency?

File: dev/sg/root/root.go [hunk 4]
```diff
@@ -14,19 +14,23 @@ var repositoryRootError error
 
 // RepositoryRoot caches and returns the value of findRoot.
 func RepositoryRoot() (string, error) {
-	once.Do(func() { repositoryRootValue, repositoryRootError = findRoot() })
+	once.Do(func() { repositoryRootValue, repositoryRootError = findRootFromCwd() })
 	return repositoryRootValue, repositoryRootError
 }
 
-// findRoot finds root path of the sourcegraph/sourcegraph repository from
+// FindRoot finds root path of the sourcegraph/sourcegraph repository from
```
Comment (efritz) [inline dev/sg/root/root.go:21]: Names need to match

File: dev/sg/output.go (hunk 3 above)
Comment (tjdevries) [reply dev/sg/output.go:15]: Yup, reversed it now.

**Review by tjdevries — COMMENTED** [review]
i don't know why i cant' cancel reviews :'(

**Review by mrnugget — APPROVED** [review]
Very nice! :clap: :clap: :clap:

Two things:

1. Can you add some basic usage examples to the [`Usage` section in the README?](https://github.com/sourcegraph/sourcegraph/tree/main/dev/sg#usage)
2. This calls for a short demo video/GIF to be posted in #dev-experience and #progress :)

**Comment by mrnugget** [comment]
(For the record: I didn't do an in-depth review since I trust you and Eric on this, but the high-level looks great to me)

**Comment by tjdevries** [comment]
Part 1, simple case (no database shenanigans)

https://user-images.githubusercontent.com/4466899/125648516-057c8900-6ad4-4b8c-b4af-08dd85a7c6de.mp4

Part 2, complicated case (much database shenanigans)

https://user-images.githubusercontent.com/4466899/125648525-8679eff3-343d-4fe5-833b-4a96467356f3.mp4

**Pull request merged**